### PR TITLE
Jetpack Boost: Add Action Endpoints

### PR DESCRIPTION
**Pull request by pyronaur**
## Proposed changes:
This adds `useDataSyncAction` and the PHP endpoint handlers, allowing to create custom endpoints that can mutate data if necessary without passing the whole state back & forth.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:
This patch is scoped to only include the package code - there's no direct implementation to test.

For implementation example, see #34599

* `jetpack_boost_register_action()` utility function: https://github.com/Automattic/jetpack/blob/49cd57ce20b8affed327e8794aaa6ccfe647bfdd/projects/plugins/boost/wp-js-data-sync.php#L32-L43

* Register PHP Endpoint: https://github.com/Automattic/jetpack/blob/49cd57ce20b8affed327e8794aaa6ccfe647bfdd/projects/plugins/boost/app/modules/image-size-analysis/data-sync/init.php#L78

* PHP Endpoint: https://github.com/Automattic/jetpack/blob/49cd57ce20b8affed327e8794aaa6ccfe647bfdd/projects/plugins/boost/app/modules/image-size-analysis/data-sync/Image_Size_Analysis_Fixer_Action.php#L12-L66

* useDataSyncAction implementation: https://github.com/Automattic/jetpack/blob/49cd57ce20b8affed327e8794aaa6ccfe647bfdd/projects/plugins/boost/app/assets/src/js/features/image-size-analysis/lib/image-fixer.ts#L22-L58

* React usage: https://github.com/Automattic/jetpack/blob/49cd57ce20b8affed327e8794aaa6ccfe647bfdd/projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/table/table.tsx#L43-L50

### CODE REVIEW

File: projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts [hunk 1]
```diff
@@ -12,7 +12,11 @@ import React from 'react';
 import { z } from 'zod';
 import { DataSync } from './DataSync';
 
-const queryClient = new QueryClient();
+/**
+ * @REACT-TODO This is temporary. We need to allow each app to define their own QueryClient.
+ * All of the functions below will have to be moved to a factory wrapper
+ */
+export const queryClient = new QueryClient();
```
Comment (thingalon) [inline projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts:19]: 👍 I'm ok with this for now but would like to this on the board to make sure we take care of it before we call the migration done.

Before I add it to the board, a quick discussion: are we sure we need to? Will it cause any harm for DataSync to have its own queryClient that it exclusively works with, and for non-datasync REact Query stuff to live alongside it? I don't see a downside.

File: projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts [hunk 2]
```diff
@@ -63,7 +70,7 @@ export function useDataSync<
 	params: Record< string, string | number > = {}
 ): DataSyncHook< Schema, Value > {
 	const datasync = new DataSync( namespace, key, schema );
-	const queryKey = [ key, ...Object.values( params ) ];
+	const queryKey = [ key, ...Object.values( params ).sort() ];
```
Comment (thingalon) [inline projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts]: 🤔 I'm not 100% sure this is right. I think the key order matters. Imagine this scenario: 

You have two parameters: `foo` and `bar`. If you're sorting by value then `{foo:1,bar:2}` will have the same query cache key as `{foo:2,bar:1}`.

I think we'd want to sort by key for consistency, then pull out the values. Something like: 

```
const queryKey = [ key, ...Object.entries.sort( ( [ a ], [ b ] ) => b-a ).map( [ k, v ] => v ) ];
```

**Review by thingalon — CHANGES_REQUESTED** [review]
One issue I see (inline), but otherwise this is great. Thank you for PRing it. :)

File: projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts (hunk 1 above)
Comment (pyronaur) [reply projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts:19]: It _may_ work but it's not correct. Think of queryClient as the namespace. Invalidation happens within a queryClient - so if I `invalidateAll`, I may accidentally invalidate applications I didn't intend to.

The fix is fairly simple - we'll just import `useDataSync` from boost, not from packages directly when we set this up - we can set those variables up in a lib directory statically. (something like `export const { useDataSync } = useDataSyncFactory()` )

File: projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts (hunk 2 above)
Comment (pyronaur) [reply projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts]: A-ha. Yes, nice catch. Fully agree.

File: projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts (hunk 1 above)
Comment (thingalon) [reply projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts:19]: 🤔 It's not a blocker here, but I don't understand. What is a scenario where we want to share a QueryClient with something that isn't DataSync?

Surely it'd be cleaner for DataSync to have its own cache, and any other use of ReactQuery to be kept separate?

File: projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts (hunk 2 above)
Comment (thingalon) [reply projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts]: Fixed in 39610c3.

**Review by thingalon — APPROVED** [review]
🚀

**Pull request merged**